BootloaderCommonPkg/Library/SmbiosInitLib/SmbiosInitLib.c

```diff
@@ -0,0 +1,422 @@
+  SmbiosEntry->EntryPointStructureChecksum  = CalculateCheckSum8 ((UINT8 *)SmbiosEntry       , sizeof (SMBIOS_TABLE_ENTRY_POINT)       );
+
+  return Status;
+}
+
+/**
+  Append an Smbios Type to an existing set of types
+  Different from AddSmbiosType, which gets called during initial Smbios setup
+
+  @param[in]  TypeData    pointer to the Type Data including strings
+  @param[in]  TypeLength  Size of type data including strings and end terminator (0000)
+
+  @retval                         EFI_DEVICE_ERROR, if Smbios Entry is NULL
+                                  Overflow status for adding this type or Type 127, otherwise
+
+**/
+EFI_STATUS
+EFIAPI
+AppendSmbiosType (
+  IN  VOID      *TypeData,
+  IN  UINT16    TypeLength
+  )
+{
+  SMBIOS_TABLE_ENTRY_POINT    *SmbiosEntry;
+  EFI_STATUS                  Status;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL || mType127Ptr == NULL || TypeLength > SMBIOS_TABLE_MAX_LENGTH) {
+    return EFI_DEVICE_ERROR;
+  }
+
+  //
+  // Check for overflows before adding new type
+  //
+  Status = CheckSmbiosOverflow (TypeLength);
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  //
+  // Find current Type127 and replace it with new Type Data
+  //
+  SmbiosEntry->TableLength -= ( ((SMBIOS_STRUCTURE *)mType127Ptr)->Length + 2);
+  CopyMem (mType127Ptr, TypeData, TypeLength);
+  SmbiosEntry->TableLength += TypeLength;
+
+  //
+  // After appending, update with Typ127 and patch entry point
+  //
+  Status = FinalizeSmbios (TypeLength);
+
+  return Status;
+}
+
+/**
+  Get the value a particular string in a Type
+  from data structure pointed to by PcdSmbiosStringsPtr
+
+  @param[in]  Type      Get the string for a Type
+  @param[in]  StringId  String Num
+
+  @retval               CHAR8 * to the string, if found
+                        "Unknown"            , otherwise
+
+**/
+CHAR8 *
+GetSmbiosString (
+  IN  UINT8   Type,
+  IN  UINT8   StringId
+  )
+{
+  SMBIOS_TYPE_STRINGS       *SmbiosStrings;
+  UINT16                    Index;
+
+  SmbiosStrings = (SMBIOS_TYPE_STRINGS *) PcdGet32 (PcdSmbiosStringsPtr);
+
+  if (SmbiosStrings == NULL) {
+    return "Unknown";
+  }
+
+  for (Index = 0; Index < PcdGet16 (PcdSmbiosStringsCnt); ++Index) {
+    if (Type == SmbiosStrings[Index].Type && StringId == SmbiosStrings[Index].Idx) {
+      return SmbiosStrings[Index].String;
+    }
+  }
+
+  return "Unknown";
+}
+
+/**
+  Add the string to an Smbios type
+
+  @param[in]  Destination     Address where the string gets added inside a type
+  @param[in]  Source          CHAR8 * to the actual string
+
+  @retval                     Destination for the next string to be added
+
+**/
+CHAR8 *
+AddSmbiosString (
+  IN  CHAR8   *Destination,
+  IN  CHAR8   *Source
+  )
+{
+  if (Source == NULL) {
+    return Destination;
+  }
+
+  CopyMem ((CHAR8 *)Destination, Source, AsciiStrLen (Source) + sizeof(CHAR8));
```

Comment:
Use AsciiStrSize(Source) to replace "AsciiStrLen (Source) + sizeof(CHAR8)"

Reply:
Thanks for pointing this out. Changed the function call to StrSize.